# Computing Agroweather analytics using Geosys Platform 

*Source: [AgroWeather Analytics](https://github.com/GEOSYS/AgroWeatherAnalytic_example) repository*

> 👋 Before reviewing this example, you must first sign-up and request your Geosys APIs credentials here :
>
> * ⚙️[Try it now](https://www.urthecast.com/geosys/geosys-api/)

> For more information on Geosys APIs :
>
> * 📚 [Geosys APIs to connect with your digital ag application](https://www.urthecast.com/geosys/geosys-api/)

> **Demo Project:** In this example, we will generate an agroweather analytic on field leveraging Geosys APIs.

### @author: Geosys

### Import Dependencies

* [pandas](https://pandas.pydata.org/) - library that we will use for loading and displaying the data in a table.

* [numpy](http://www.numpy.org/) - library that we will use for linear algebra operations.

* [geopandas](https://pypi.org/project/geopandas/) - Python module that is built on top of Pandas extending its functionalities to works with spatial data.

* [datetime](https://docs.python.org/3/library/datetime.html) - library that we will use for manipulating dates and time in a simple way.

* [request](https://pypi.org/project/requests/) - library that we will make requests by using the HTTP protocol more easier.

In [1]:
import os.path as pa
import sys
import os
from os import listdir
from os.path import isfile, join
import requests
import requests.exceptions as rex
import datetime as dt

import pandas as pd
import geopandas as gpd
import numpy as np
import getpass
import json


### Authentication on the Geosys APIs 

To connect and use Geosys API on the staging environment, please sure you have API credentials : username and password. 

> Sign up and request here :⚙️[Try it now](https://www.urthecast.com/geosys/geosys-api/)  

The authentification is based on **the OAuth 2.0 Password Grant Type** with a bearer token. 

⚠️ This token access is valid for an hour. After expiration, you will need to request another token.


In [2]:
username = getpass.getpass()

········


In [3]:
password = getpass.getpass()

········


In [4]:
Authentication_url='https://identity.preprod.geosys-na.com/v2.1/connect/token'
response=requests.post(Authentication_url, data={'grant_type':'password','scope':'openid',
                         'username':username,'password':password},
                          headers={'Authorization':'Basic c3dhZ2dlcjpzd2FnZ2VyLnNlY3JldA==',
                             'Accept':'application/json, text/plain, */*',
                             'Content-Type':'application/x-www-form-urlencoded'})
result=response.json()

access_token=result['access_token']
print(" Your access token : ",access_token)

 Your access token :  eyJhbGciOiJSUzI1NiIsImtpZCI6IjMyNUJCRDNEOTZCODAxRkFBODRBQUU5MzhBQkI4Njk1RDJBNUIyN0RSUzI1NiIsInR5cCI6IkpXVCIsIng1dCI6Ik1sdTlQWmE0QWZxb1NxNlRpcnVHbGRLbHNuMCJ9.eyJuYmYiOjE2MTkyMDgzMjMsImV4cCI6MTYxOTIxMTkyMywiaXNzIjoiaHR0cHM6Ly9pZGVudGl0eS5wcmVwcm9kLmdlb3N5cy1uYS5jb20vdjIuMSIsImF1ZCI6Imh0dHBzOi8vaWRlbnRpdHkucHJlcHJvZC5nZW9zeXMtbmEuY29tL3YyLjEvcmVzb3VyY2VzIiwiY2xpZW50X2lkIjoic3dhZ2dlciIsInN1YiI6IjEwMDExMjk3MiIsImF1dGhfdGltZSI6MTYxOTIwODMyMywiaWRwIjoibG9jYWwiLCJnZW82X3N1YiI6IjJSNklPRG9RZ1R1SFZ5aWJXSjFKRksiLCJpYXQiOjE2MTkyMDgzMjMsInNjb3BlIjpbIm9wZW5pZCJdLCJhbXIiOlsicGFzc3dvcmQiXX0.KmKgK9LD3QHvNTHpBD1CcQRvs7945dGRfWs6g63QsGGrEHVQDZR2m9zmftiLy2alVI6MB_FIAV0VPvCQY2KGj6sfqZzeJuudalIr3fCJxDgSbW1CuQgAXKtRO4j8d6QbwZ6usuiVI1L1yXBI_d-OnebVL_f3YElxTeYEanxVur_YUcGNlTdcUJTEDMU-Fkk7v_8rc_pJFcBnDQelO7bTPjkQPHDegmYpBKsoTRu4VlpB52sdtLiITGY0spFpMtve4UThZ0s2nW0OqyH0XuaoEKvH6pck7TZ2YcWdQlmCoiJ8twsXa9eml2qC3ZYEnZE54iOsnn8y0ZxZwc-0GaBigQ


### Data loading

In this demo we will use a dataset containing field border in .csv format.

📈 The dataset contains: field name, sowing date, crop ID (the current crop on the field) and the geometry, as WKT polygon. Please update this file to include your field of interest.

> data set format
>
> * `Name` String object, can be capitalized or lowercase letters
>
> * `Sowing_date` Datetime format  (ex : 2021/01/01)
>
> * `crop ID` String object, must be **capital** letters
>
> * `WKT` Vector geometric object. It is a standard format in text mode used to represent a geometry (ex : POLYGON ((55.876264 48.838602, 55.877344 48.852597, 55.874162 48.855574000000004, 55.874253 48.840364, 55.876264 48.838602)))

In [5]:
# Load the data.
input_file = pd.read_csv('Field.csv',sep=';')

# Print the data table.
input_file.head(10)


Name sowing_date crop_id  \
0  Field_1  15/06/2019    CORN   
1  Field_2  01/05/2020    CORN   
2  Field_3  15/05/2020    CORN   

                                                 WKT  
0  POLYGON((-86.86701653386694 41.331532756357426...  
1  POLYGON ((55.876264 48.838602, 55.877344 48.85...  
2  POLYGON ((55.982660360000004 49.32906512, 55.9...

### Create fields and get centroid

⛅ Before calling the Geosys Weather service API, we need to create field and get their centroid. This can be achieved using the business entity service  **http://api-pp.geosys-na.net/master-data-management/v6/**.

Response body examples:

> {
> "id": "15w9xj2"
> }

> {
> "errors": {
> "body": {
> "sowingDate": \[
> {
> "code": "crop_cycle",
> "message": "A season field with a sowing date too close already exists. Id: 65w2xnn, SowingDate: >03/01/2021 00:00:00"
> }
> \]
> }
> },
> "code": "model_validation_error",
> "message": "Bad request: see the Errors field for details"
> }
>
> ⚠️ This type of response body does not mean that the API call did not work but indicates that a season field has been already created. Into the code we handle this error.

In [6]:
len(input_file)
weather_centroid=[]
sowing_date=0
create_season_field_url = "http://api-pp.geosys-na.net/master-data-management/v6/seasonfields"
headers={'Authorization':'Bearer '+access_token,
                                        'Accept':'application/json','Content-Type': 'application/json'
                                        }

for idx in range(0,len(input_file)):
    sowing_date = pd.to_datetime(input_file['sowing_date'][idx])    
    payload = json.dumps({"Geometry":input_file['WKT'][idx],"Crop": {"Id": input_file['crop_id'][idx]},"SowingDate": sowing_date.strftime('%Y-%m-%d')})
    response = requests.request("POST", create_season_field_url, headers=headers, data=payload)
    data=response.json()
    if response.status_code == 400:
        season_field_id_weather=data['errors']['body']['sowingDate'][0]['message'].split('Id:')
        season_field_id_weather=season_field_id_weather[1][1:8]
    if response.status_code == 201:
        season_field_id_weather=data['id']    
    get_centroid_url = "http://api-pp.geosys-na.net/master-data-management/v6/seasonfields/"+season_field_id_weather
    response = requests.request("GET", get_centroid_url, headers=headers)
    data=response.json()
    centroid=data['centroid']

    weather_centroid.append(centroid)
print("The fields centroid are : ", weather_centroid)

The fields centroid are :  ['POINT (-86.86482785 41.328020020000004)', 'POINT (55.87552231 48.84733722)', 'POINT (55.98288595 49.33248007)']


### Initialize variable for our analytic

Analytics we want to compute are an integration of weather metric over a period. This allows to assess the risk of specific pest devolpment and therefore the interest to consider preventive actions.\n\nTo initialize our processing, we will define our integration period. The current date is used as the default value to determine period dates.


In [7]:
today = dt.date.today()
end = today-dt.timedelta(days=1)
start_5=today-dt.timedelta(days=6)
start_10=today-dt.timedelta(days=11)
start_15=today-dt.timedelta(days=16)

periode=[start_5,start_10,start_15]


### Weater data

⛅ For each periode, we will use Geosys Weather Service to fetch historical weather metric parameters like the temperature and the cumulative rainfall on our fields.

Example of response body for a date and a point location (field centroid):

> {'location': 'POINT (-86.86482785 41.328020020000004)', 'date': '2021-04-17T00:00:00Z', 'temperature': {'standard': 42.26000000619203}, 'precipitation': {'cumulative': 0.0}}, {'location': 'POINT (-86.86482785 41.328020020000004)', 'date': '2021-04-18T00:00:00Z', 'temperature': {'standard': 49.262000000590405}, 'precipitation': {'cumulative': 0.0}}, {'location': 'POINT (-86.86482785 41.328020020000004)', 'date': '2021-04-19T00:00:00Z', 'temperature': {'standard': 44.906000004075175}, 'precipitation': {'cumulative': 0.15748031496062992}}, {'location': 'POINT (-86.86482785 41.328020020000004)', 'date': '2021-04-20T00:00:00Z', 'temperature': {'standard': 34.790000012167965}, 'precipitation': {'cumulative': 0.2559055118110236}}, {'location': 'POINT (-86.86482785 41.328020020000004)', 'date': '2021-04-21T00:00:00Z', 'temperature': {'standard': 34.95200001203842}, 'precipitation': {'cumulative': 0.1653543307086614}}

In [8]:
weather_data=[]
for p in periode:
    print ("periode time : (",p, "-",end,")")
    sortie=[]
    for idx in range(0,len(weather_centroid)):
        url_weather='http://api-pp.geosys-na.net/Weather/v1/weather?'\
                '%24offset=0&%24limit=2000&%24count=false'\
                '&Location='+weather_centroid[idx]+\
                '&Date=$between:'+p.strftime('%Y-%m-%d')+'T00:00:00.0000000Z|'+end.strftime('%Y-%m-%d')+'T00:00:00.0000000Z' \
                '&Provider=GLOBAL1'\
                '&WeatherType=HISTORICAL_DAILY'\
                '&Temperature.Standard=$gte:0'\
                '&$fields=Location,Date,Temperature.Standard,Precipitation.Cumulative'
        
        #  Requetage de l'API
        response=requests.get(url_weather, 
                                headers={'Authorization':'Bearer '+access_token,
                                        'Accept':'application/json, text/plain, */*'
                                        })
        data=response.json()
        for j in data:
            j['Name']=input_file['Name'][idx]
            j['Geometry']=input_file['WKT'][idx]
            j['Periode']= (end-p).days
            
        
        
        sortie.append(data)
    weather_data.append(sortie)

periode time : ( 2021-04-17 - 2021-04-22 )
periode time : ( 2021-04-12 - 2021-04-22 )
periode time : ( 2021-04-07 - 2021-04-22 )


### Create the output structure

We create here a dataset to save analytics.

The output dataset contains for the 5, 10 and 15 days integratin period:

. ☀️ The sum of temperatures
. ☔ the sum of rainfalls
. the response the number of days where the rainfall are greater than 2 milimiters

In [9]:
output_data = pd.DataFrame(columns=['Name',
                                    'Geometry',
                                    'centroid',
                                    'Sum of Precip (mm) 5 days',
                                    'Rain days>2mm 5 days',
                                    'Sum of Tmean (C) 5 days',
                                    'Sum of Precip (mm) 10 days',
                                    'Rain days>2mm 10 days',
                                    'Sum of Tmean (C) 10 days',
                                    'Sum of Precip (mm) 15 days',
                                    'Rain days>2mm 15 days',
                                    'Sum of Tmean (C) 15 days'
                                    ])

### Compute AgroWeather analytics 

In the current section, we will use weather metric to compute our AgroWeahter analytic by integrating daily weather metrics on each field.


In [10]:
output_data=[]
res=[]
for s2 in weather_data:
    output_data=[]
    for s in s2:
        rain_day=0
        df=pd.DataFrame(s)
        tmean= np.array(df.apply(lambda x: x['temperature']['standard'],axis=1))
        location= np.array(df.apply(lambda x: x['location'],axis=1))
        wkt= np.array(df.apply(lambda x: x['Geometry'],axis=1))
        field= np.array(df.apply(lambda x: x['Name'],axis=1))
        d= np.array(df.apply(lambda x: x['date'],axis=1))
        rr= np.array(df.apply(lambda x: x['precipitation']['cumulative'],axis=1))
        periode=df['Periode']
        
        # Sum of precip
        sumofrr=0
        for l in rr:
            sumofrr += l                   

        # Sum of Tmean
        sumoftmean=0
        for t in tmean:
            sumoftmean += t                   

        # Days with Rain >2mm
        rain_day = 0
        for j in rr:
            if j>2:
                rain_day=rain_day+1

        output_data.append({'Name': field[0],
                            'Geometry': wkt[0],
                            'centroid':location[0],
                            'Sum of Precip (mm) %s days'%(periode[0]): sumofrr,
                            'Rain days>2mm %s days'%(periode[0]):rain_day,
                            'Sum of Tmean (F) %s days'%(periode[0]):sumoftmean,
                                })
    
    output_data=pd.DataFrame(output_data)
    res.append(output_data)

res2 = [df.set_index(['Name','Geometry']) for df in res]
final_df = pd.concat(res2,axis=1)
# print(final_df)


## Analytic extraction

AgroWeather analytics will be exported as a csv file in the notebook folder.

In [11]:
final_df.to_csv('Geosys_weather_monitoring_'+end.strftime('%Y-%m-%d')+'.csv',sep=';',index=False)